unittests/test_typecheck.py

```diff
@@ -251,3 +251,42 @@ def __init__(self, x, y):
 
     assert Y('1').y == 1
     assert Z(5, 3).z  == 8
+
+
+def test_composition_of_types():
```

Comment:
Is this list of tests complete or exhaustive? Should it be?

Reply:
What do you think it's missing? Perhaps I could add a composition for Dicts and Tuples. Is it something else?

Reply:
Dicts were in my mind, but I didn't want to be explicit because I was unsure if it was the only missing type to add... 
Tuples, as you have said, are also missing... Do we have any other type that we support?
If so, we should add it here... What do you think?

Reply:
I've added more unit tests and it actually caught an error in the implementation.